# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/canvas/CanvasRenderingContext2D.cpp
# Commit: 88ac974e3453
# Full Hash: 88ac974e3453cb29f1c946bbb044654b18c4c973
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-19 03:56:45
# Description:
#   Bug 1558009: Always return false from CanvasRenderingContext2D::EnsureTarget when mTarget == sErrorTarget. r=mattwoodrow
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D34922
# ==============================================================================

diff -r 9e8b490acc3b -r 88ac974e3453 dom/canvas/CanvasRenderingContext2D.cpp
--- a/dom/canvas/CanvasRenderingContext2D.cpp	Tue Jun 18 18:43:05 2019 +0000
+++ b/dom/canvas/CanvasRenderingContext2D.cpp	Tue Jun 18 22:10:47 2019 +0000
@@ -1227,7 +1227,7 @@
   }
 
   if (mTarget) {
-    return true;
+    return mTarget != sErrorTarget;
   }
 
   // Check that the dimensions are sane
@@ -1666,9 +1666,16 @@
 
 already_AddRefed<mozilla::gfx::SourceSurface>
 CanvasRenderingContext2D::GetSurfaceSnapshot(gfxAlphaType* aOutAlphaType) {
+  if (aOutAlphaType) {
+    *aOutAlphaType = (mOpaque ? gfxAlphaType::Opaque : gfxAlphaType::Premult);
+  }
+
   if (!mBufferProvider) {
     if (!EnsureTarget()) {
-      return nullptr;
+      MOZ_ASSERT(
+          mTarget == sErrorTarget,
+          "On EnsureTarget failure mTarget should be set to sErrorTarget.");
+      return mTarget->Snapshot();
     }
   }
 
@@ -1680,10 +1687,6 @@
   RefPtr<DataSourceSurface> dataSurface = snapshot->GetDataSurface();
   mBufferProvider->ReturnSnapshot(snapshot.forget());
 
-  if (aOutAlphaType) {
-    *aOutAlphaType = (mOpaque ? gfxAlphaType::Opaque : gfxAlphaType::Premult);
-  }
-
   return dataSurface.forget();
 }
 
@@ -4937,7 +4940,7 @@
 
   if (!mBufferProvider) {
     if (!EnsureTarget()) {
-      return NS_ERROR_OUT_OF_MEMORY;
+      return NS_ERROR_FAILURE;
     }
   }
 
